src/Compilers/Extension/SetGlobalProjectCollectionForCPS.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem;
+using Microsoft.VisualStudio.ProjectSystem.Build;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Roslyn.Compilers.Extension
+{
+    [ExportBuildGlobalPropertiesProvider]
+    [AppliesTo("(" + ProjectCapabilities.CSharp + " | " + ProjectCapabilities.VB + ")" + " & " + ProjectCapabilities.LanguageService)]
+    public class SetGlobalGlobalPropertiesForCPS : StaticGlobalPropertiesProviderBase
+    {
+        [ImportingConstructor]
+        public SetGlobalGlobalPropertiesForCPS(IProjectService projectService, ILocalRegistry2 registry, IProjectThreadingService threadingService)
+            : base(projectService.Services)
+        {
+            Registry = registry;
+            ThreadingService = threadingService;
+        }
+
+        public ILocalRegistry2 Registry { get; }
+        public IProjectThreadingService ThreadingService { get; }
+
+        public override async Task<IImmutableDictionary<string, string>> GetGlobalPropertiesAsync(CancellationToken cancellationToken)
+        {
+            await ThreadingService.SwitchToUIThread();
+
+            Registry.GetLocalRegistryRoot(out var localRegistryRoot);
+            var registryParts = localRegistryRoot.Split('\\');
```

Comment:
Is it a requirement that the registry be used on the UI thread?

Reply:
Most VS APIs are UI-thread unless explicitly documented.

Reply:
I'll just compute it on package load so we don't need to have any JTF in roslyn :)

Reply:
Thank you :smile: